### DISCUSSION TITLE
brew install rust (error installing rust)

### DISCUSSION THREAD

**Discussion by 2246sam20** [Homebrew/discussions#1848] [Getting started]
Hello!
I m trying to install EFL but when brew tries to install rust one of the dependencies of EFL , it throws an error.
 Does anyone know why this might be happening?
A help will be much appreciated.

**Comment by gromgit** [comment]
To avoid having to make wild-ass guesses, please post:
* The actual error you saw
* The output of `brew config` and `brew doctor`.

**Reply by 2246sam20** [reply]
Sorry for that actually when i was posting the problem brew install rust was in progress to capture the error and it takes too much time and in the end it just throws an error.
My Mac OS Verson - 10.13
**Below is the error:**
```

==> Installing rust

==> ./configure --prefix=/usr/local/Cellar/rust/1.53.0 --release-channel=stable
==> make
Last 15 lines from /Users/admin/Library/Logs/Homebrew/rust/02.make:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:325:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
13 errors generated.
[8/2716] Building CXX object lib/Demangle/CMakeFiles/LLVMDemangle.dir/ItaniumDemangle.cpp.o
ninja: build stopped: subcommand failed.
thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', /Users/admin/Library/Caches/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.44/src/lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
	finished in 136.297 seconds
failed to run: /private/tmp/rust-20210719-37826-1nq149y/rustc-1.53.0-src/build/bootstrap/debug/bootstrap build --stage 2
Build completed unsuccessfully in 0:13:44
make: *** [all] Error 1
```


**Output of brew config**

```
HOMEBREW_VERSION: 3.2.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7d37d09709b6fbd5b89b468eb0b4438ffa0fb81c
Last commit: 8 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c499a4d0364409884d9a828a9f3229b60b30219c
Core tap last commit: 19 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.FrsFfLpxb2/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit sandybridge
Clang: 10.0.0 build 1000
Git: 2.17.2 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.13.6-x86_64
CLT: 10.1.0.0.1.1539992718
Xcode: 10.1
XQuartz: 2.7.11 => /opt/X11
```

**Output of brew doctor:**

```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /opt/local/bin/freetype-config
  /opt/local/bin/icu-config
  /opt/local/bin/libpng-config
  /opt/local/bin/libpng16-config
  /opt/local/bin/ncurses6-config
  /opt/local/bin/ncursesw6-config
  /opt/local/bin/pcre-config
  /opt/local/bin/python3.7-config
  /opt/local/bin/python3.7m-config
  /opt/local/bin/sdl2-config
  /opt/local/bin/uuid-config
  /opt/local/bin/xml2-config

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libtcl8.6.dylib
  /usr/local/lib/libtk8.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/fakemysql.h
  /usr/local/include/fakepq.h
  /usr/local/include/fakesql.h
  /usr/local/include/itcl.h
  /usr/local/include/itcl2TclOO.h
  /usr/local/include/itclDecls.h

**NOTE - I have removed the remaining lines because it was too large**

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/tcl.pc
  /usr/local/lib/pkgconfig/tk.pc


Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libtclstub8.6.a
  /usr/local/lib/libtkstub8.6.a

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.9

Warning: You are using macOS 10.13.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.


Warning: /usr/bin occurs before /usr/local/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/usr/local/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/usr/local/bin:$PATH"' >> /Users/admin/.bash_profile

The following tools exist at both paths:
  info
  infokey
  install-info
  makeinfo
  texi2dvi
  texi2pdf
  texindex

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> /Users/admin/.bash_profile

Warning: Broken symlinks were found. Remove them with `brew cleanup`:
  /usr/local/bin/R
  /usr/local/bin/Rscript
  /usr/local/bin/code
  /usr/local/bin/prl_convert
  /usr/local/bin/prl_disk_tool
  /usr/local/bin/prl_perf_ctl
  /usr/local/bin/prlcore2dmp
  /usr/local/bin/prlctl
  /usr/local/bin/prlexec
  /usr/local/bin/prlsrvctl
  /usr/local/share/man/man8/prl_convert.8
  /usr/local/share/man/man8/prl_disk_tool.8
  /usr/local/share/man/man8/prlctl.8
  /usr/local/share/man/man8/prlexec.8
  /usr/local/share/man/man8/prlsrvctl.8
  /usr/local/share/zsh/site-functions/_brew_cask
```

**Reply by carlocab** [reply]
You're on your own here, I'm afraid. See:
```
Warning: You are using macOS 10.13.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.
```

If you just need Rust, I recommend using [`rustup`](https://rustup.rs). If you need Rust to install another formula, this might work:
```
brew install llvm
brew install --cc=llvm_clang rust
```
Note that you will need to build `llvm` from source, and then build `rust` from source. This will take a very long time. If you're using a laptop, I suggest you plug it into a power supply and turn off your energy saver settings to avoid it going to sleep while compiling.

**Reply by 2246sam20** [reply]
I want rust to install efl.
OK will try the commands you mentioned.

**Reply by 2246sam20** [reply]
hello @carlocab 
So llvm has been installed
```
brew info llvm:
dmins-MacBook-Pro:~ admin$ brew info llvm
llvm: stable 12.0.1, HEAD [keg-only]
Next-gen compiler infrastructure
https://llvm.org/
/usr/local/Cellar/llvm/12.0.1 (9,857 files, 1.6GB)
  Built from source on 2021-07-21 at 09:24:59
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/llvm.rb
License: Apache-2.0 with LLVM-exception
==> Dependencies
Build: cmake ✔, swig ✔
Required: python@3.9 ✔, libffi ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
To use the bundled libc++ please add the following LDFLAGS:
  LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib"

llvm is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have llvm first in your PATH, run:
  echo 'export PATH="/usr/local/opt/llvm/bin:$PATH"' >> /Users/admin/.bash_profile

For compilers to find llvm you may need to set:
  export LDFLAGS="-L/usr/local/opt/llvm/lib"
  export CPPFLAGS="-I/usr/local/opt/llvm/include"

==> Analytics
install: 25,172 (30 days), 69,351 (90 days), 258,816 (365 days)
install-on-request: 16,652 (30 days), 50,034 (90 days), 200,076 (365 days)
build-error: 0 (30 days)
```


BUT when i try the command `brew install --cc=llvm_clang rust` 
Its throwing the error:

```
==> ./configure --prefix=/usr/local/Cellar/rust/1.53.0 --release-channel=stable
==> make
Last 15 lines from /Users/admin/Library/Logs/Homebrew/rust/02.make:
   Compiling globset v0.4.5
   Compiling ignore v0.4.17
   Compiling merge_derive v0.1.0
   Compiling merge v0.1.0
   Compiling toml v0.5.7
    Finished dev [unoptimized + debuginfo] target(s) in 3m 40s
thread 'main' panicked at '

couldn't find required command: "llvm_ar"

', src/bootstrap/sanity.rs:60:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
failed to run: /private/tmp/rust-20210721-24052-14ikjba/rustc-1.53.0-src/build/bootstrap/debug/bootstrap build --stage 2
Build completed unsuccessfully in 0:04:35
make: *** [all] Error 1

Do not report this issue to Homebrew/brew or Homebrew/core!

These open issues may also help:
Rust-dependent formulae on Apple Silicon - upstream issue tracker https://github.com/Homebrew/homebrew-core/issues/68301
```
So it couldnt find `couldn't find required command: "llvm_ar"`
So how to resolve this?
Thank you

**Reply by carlocab** [reply]
Try adding this to the formula:
```diff
diff --git a/Formula/rust.rb b/Formula/rust.rb
index 253301d8dc..c83071dda6 100644
--- a/Formula/rust.rb
+++ b/Formula/rust.rb
@@ -34,6 +34,7 @@ class Rust < Formula
   depends_on "ninja" => :build
   depends_on "python@3.9" => :build
   depends_on "libssh2"
+  depends_on "llvm"
   depends_on "openssl@1.1"
   depends_on "pkg-config"
```
Note that (if this works) this will likely create linkage between `rust` and `llvm`, so you'll need to keep `llvm` installed if you want Rust to work.

**Reply by 2246sam20** [reply]
so when you say this Try adding this to the formula:
Do you mean to add this line : 
`diff --git a/Formula/rust.rb b/Formula/rust.rb`

so the final formula : 
`brew install --cc=llvm_clang rust diff --git a/Formula/rust.rb b/Formula/rust.rb`

**Reply by carlocab** [reply]
No, sorry. That's just a commonly used format for a patch. What it actually means is that you're meant to add
```ruby
depends_on "llvm"
```
to `rust.rb` in between the lines shown on my previous comment. You can edit your local copy of `rust.rb` with `brew edit rust`. You may need to set your terminal's `EDITOR` (or otherwise be comfortable with `vim`, as that is what it will fall back to).

**Reply by carlocab** [reply]
Alternatively, you can copy the content of my patch (the GitHub UI has a copy button on the top right of those textboxes), and then do
```
cd $(brew --repo homebrew/core)
pbpaste | patch -p1
```
to apply the change I suggested to `rust.rb`.

**Reply by 2246sam20** [reply]
ok thanks  for the clear and quick reply
Will make the changes.

**Reply by 2246sam20** [reply]
Hello! @carlocab 
still the same error 
`https://github.com/Homebrew/discussions/discussions/1848#discussioncomment-1030324`

**Reply by Remigiuszlg** [reply]
> You're on your own here, I'm afraid. See:
> 
> ```
> Warning: You are using macOS 10.13.
> We (and Apple) do not provide support for this old version.
> You will encounter build failures with some formulae.
> Please create pull requests instead of asking for help on Homebrew's GitHub,
> Twitter or any other official channels. You are responsible for resolving
> any issues you experience while you are running this
> old version.
> ```
> 
> If you just need Rust, I recommend using [`rustup`](https://rustup.rs). If you need Rust to install another formula, this might work:
> 
> ```
> brew install llvm
> brew install --cc=llvm_clang rust
> ```
> 
> Note that you will need to build `llvm` from source, and then build `rust` from source. This will take a very long time. If you're using a laptop, I suggest you plug it into a power supply and turn off your energy saver settings to avoid it going to sleep while compiling.

**Comment by SMillerDev** [comment]
It's impossible to know without seeing the error.

**Reply by 2246sam20** [reply]
Hello!
I have mentioned the error along with the other details in the above reply.
Below is the link for the same.
https://github.com/Homebrew/discussions/discussions/1848#discussioncomment-1024658

Thank You.

**Comment by jspraul** [comment]
"The issue is likely that the cmath file is including the wrong version of math.h based on the search order of the header files"
-- https://stackoverflow.com/questions/58628377

Still looking into which of the answers there might solve it for me, I think I need more info on what homebrew does to manage the search order of the header files.

Also look at https://github.com/Homebrew/discussions/discussions/969#discussioncomment-418035, two potential ways to work around a dependency on rust after installing rust using https://rustup.rs.

**Comment by frakman1** [comment]
I ran into similar rust and dependency installation problems that stemmed from some cmath related issue.

I just wanted to report that I was successfull in getting rust to upgrade from 1.47.0 -> 1.55.0 on High Sierra (MacOS 10.13.6).

I had to manipulate the file:

`/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath`

a couple of times. Once to get past an earlier brew failure for a different module. 
If I recall correctly, that module was openblas, and I had to also modify `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openblas.rb` to add a make argument as such:
```
-    system "make", "CC=#{ENV.cc}", "FC=gfortran", "libs", "netlib", "shared"
+    system "make", "CC=#{ENV.cc}", "FC=gfortran", "libs", "netlib", "shared", "NO_AVX512=1"
```

The other change was to replace:

`#include <math.h>`

with

`#include</Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/math.h>`

But that later caused rust compilation to fail mysteriously, so I changed it back to get it to work.

While troubleshooting, I did the following in an attempt to compile it manually and insert environment variables to produce more verbose output.
I looked at `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/rust.rb` to see what the compilations steps were in the `install` function

Go to directory: `~/Library/Caches/Homebrew/downloads`

Untar the file xxxxx--rustc-1.55.0-src.tar.gz

Then navigated to `~/Library/Caches/Homebrew/downloads/rustc-1.55.0-src`
ran: `./configure --prefix=/usr/local/Cellar/rust/1.55.0 --release-channel=stable` (just like `rust.rb` instructs)
`RUST_BACKTRACE=full`  (for more verbose output. You can also set it to 1)
You can also do this in the `rust.rb` file in the `install` function using: 
`ENV["RUST_BACKTRACE"] = "full"` 
then run `make` 
(several iterations later)
`make install`

I probably missed out on the other post install instructions in the `rust.rb` formula but at least brew now thinks rust is installed for other dependencies to work.

Result:
```bash
$brew info rust
rust: stable 1.55.0, HEAD
Safe, concurrent, practical language
https://www.rust-lang.org/
/usr/local/Cellar/rust/1.47.0 (12,057 files, 580.6MB)
  Poured from bottle on 2021-02-14 at 21:34:26
/usr/local/Cellar/rust/1.55.0 (29,361 files, 704.3MB) *
  Built from source
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/rust.rb
License: Apache-2.0 or MIT
==> Dependencies
Build: cmake ✔, ninja ✔, python@3.9 ✔
Required: libssh2 ✔, openssl@1.1 ✔, pkg-config ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 38,863 (30 days), 91,061 (90 days), 270,176 (365 days)
install-on-request: 24,594 (30 days), 47,565 (90 days), 148,019 (365 days)
build-error: 0 (30 days)
```


Now qt upgrade 5.15.2 -> 6.1.3 is failing but that's [another story](https://github.com/Homebrew/discussions/discussions/2157).

**Comment by cang-mang** [comment]
Linuxbrew, installed rust 1.62.0 x86_64.
I have this problem:

```
$rustc --version
rustc: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
```

```
$cargo --version
cargo: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
```

**Reply by gromgit** [reply]
What's the output of `brew config` and `brew doctor` on your system?

**Comment by xlla** [comment]
I'v install rust@1.74 with llvm@15 on macos 10.13.
When I try to install rust@1.83 with llvm@18, it failure on link rustc_driver.dylib

<pre>
 "-o" "/private/tmp/rust-20260228-44768-s8aiyx/rustc-1.83.0-src/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc_driver-f95b5e658c91b7ee.dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-dylib" "-Wl,-install_name" "-Wl,@rpath/librustc_driver-f95b5e658c91b7ee.dylib" "-nodefaultlibs" "-Wl,-rpath,@loader_path/../lib"
  = note: Undefined symbols for architecture x86_64:
            "vtable for std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >", referenced from:
                __ZNSt3__119basic_ostringstreamIcNS_11char_traitsIcEENS_9allocatorIcEEEC1B8ne180100Ev in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
            NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
            "std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >::str() const", referenced from:
                _LLVMRustPrintTargetCPUs in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
            "VTT for std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >", referenced from:
                _LLVMRustPrintTargetCPUs in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
                __ZNSt3__119basic_ostringstreamIcNS_11char_traitsIcEENS_9allocatorIcEEEC1B8ne180100Ev in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
                std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_ostringstream() in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
            "vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >", referenced from:
                _LLVMRustPrintTargetCPUs in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
                __ZNSt3__119basic_ostringstreamIcNS_11char_traitsIcEENS_9allocatorIcEEEC1B8ne180100Ev in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
                std::__1::basic_ostringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_ostringstream() in librustc_codegen_llvm-ca4c576b704b5d25.rlib(7372e21ddc8be4fd-PassWrapper.o)
            NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: could not compile `rustc_driver` (lib) due to 1 previous error


</pre>

Who can share the experiments of build recently new rust on this ancient macos :)
